src/utils/archive.rs

```diff
@@ -0,0 +1,1591 @@
+    /// archive is broken or cannot be accessed.
+    pub fn delete(
+        &mut self,
+        name: &[u8],
+        check: impl FnOnce(&Meta) -> Result<(), Meta::ConsistencyError>,
+    ) -> Result<(), AccessError<Meta::ConsistencyError>> {
+        let hash = self.hash_name(name);
+        let found = match self.find(hash, name)? {
+            Some(found) => found,
+            None => return Err(AccessError::NotFound),
+        };
+        check(
+            &self.file.read(found.meta_start(), |read| Meta::read(read))?
+        ).map_err(AccessError::Inconsistent)?;
+        Ok(self.delete_found(hash, found)?)
+    }
+
+    /// Deletes an object after it has been found.
+    ///
+    /// This unlinks the object from its bucket chain and replaces it with an
+    /// empty object.
+    fn delete_found(
+        &mut self, hash: u64, found: FoundObject
+    ) -> Result<(), ArchiveError> {
+        match found.prev {
+            Some(pos) => {
+                ObjectHeader::update_next(
+                    pos.into(), found.header.next, &mut self.file)?
+            }
+            None => self.set_index(hash, found.header.next)?,
+        }
+        self.create_empty(found.start, found.header.size)?;
+        Ok(())
+    }
+
+    /// Creates an empty object.
+    fn create_empty(
+        &mut self, start: u64, mut size: u64
+    ) -> Result<(), ArchiveError> {
+        let next_start = start.saturating_add(size);
+        if next_start < self.file.size {
+            let header = ObjectHeader::read(&self.file, next_start)?;
+            if header.name_len.is_none() {
+                self.unlink_empty(next_start, header.next)?;
+                size += header.size;
+            }
+        }
+        ObjectHeader::new_empty(size, self.get_empty_index()?).write(
+            &mut self.file, start
+        )?;
+        self.set_empty_index(NonZeroU64::new(start))?;
+        Ok(())
+    }
+
+    /// Unlinks an empty object from the empty chain.
+    fn unlink_empty(
+        &mut self, start: u64, next: Option<NonZeroU64>
+    ) -> Result<(), ArchiveError> {
+        let mut curr = self.get_empty_index()?;
+        let start = NonZeroU64::new(start);
+
+        // We are the start of the chain.
+        if curr == start {
+            self.set_empty_index(next)?;
+            return Ok(())
+        }
+
+        // We are further down the chain.
+        while let Some(pos) = curr {
+            let header = ObjectHeader::read(&self.file, pos.into())?;
+            if header.next == start {
+                ObjectHeader::update_next(pos.into(), next, &mut self.file)?;
+                return Ok(())
+            }
+            curr = header.next;
+        }
+
+        // We are not in the chain at all???
+        Err(ArchiveError::Corrupt)
+    }
+
+    /// Finds the start of the object with the given name.
+    ///
+    /// The returned position is right after the name.
+    fn find(
```

Comment:
So, there is a sequential scan whenever `find` is called, right? While this works, I wonder if a table of names to starting index would be useful (of course, it would have to be a bit smart about growth).

Reply:
Because this is a hash map, the number of entries in a chain of each bucket should be really small. This can be optimized by choosing a good number of buckets. In most cases the file will be memory mapped, too, so there shouldn’t be actual reads either.

I _think_ this approach will be good enough, but we’ll have to see in practice.